### ISSUE TITLE
Upload arbitrary files?

### ISSUE THREAD

**Issue by renpytom** [itchio/butler#45]
Is there a way to upload an arbitrary file (like an .apk or a .tar.bz2) to a channel using butler? Even if it doesn't get the nice speed boost from the compression, having such a feature would be nice, as it would allow tools (such as Ren'Py) to use butler for all upload needs.

**Comment by fasterthanlime** [comment]
@renpytom currently you'll have to create a directory with the .apk file in there and upload that, for example.

For .tar.gz / .tar.bz2, are you trying to distribute Linux builds? The backend generates .zips for now but there's no reason it couldn't be configured to also generate other types of archives.

**Comment by renpytom** [comment]
This doesn't seem to be working. When I upload a directory with a single file - a .tar.bz2 or a .apk - it seems to fail to show up in the interface. 

(This is with the renpytom/the-question project.)

I'm a little worried with uploading a directory. Windows can't cope with the executable bits found in our -mac and -all zips, and in Linux tarballs. So I try not to write anything but the archive file to disk.

**Comment by fasterthanlime** [comment]
Strange. The logs show:

> Failed processing task_d56ded6e-7222-48d9-8fd7-736ff679f0f4. Error = while applying patch: the_question-1.0-linux.tar.bz2: expected to write 16507790 bytes, wrote 16507786 bytes

and

> Failed processing task_da9fc38d-b7cf-40e0-a566-6f3495b4dc8b. Error = while applying patch: MoonlightWalks-2.1.2-release.apk: expected to write 17561613 bytes, wrote 17561609 bytes

Does doing `butler diff /dev/null folder-with-archive` and `butler apply patch.pwr --signature patch.pwr.sig /dev/null -d stage` work? (And does `stage` contain a valid .tar.bz2 / .apk after)

**Comment by fasterthanlime** [comment]
> I'm a little worried with uploading a directory. Windows can't cope with the executable bits found in our -mac and -all zips, and in Linux tarballs. So I try not to write anything but the archive file to disk.

butler actually detects & fixes permissions for Linux & Mac binaries, so that's not a problem :) Several people upload directories from Windows, which indeed lacks the proper permission support.

**Comment by renpytom** [comment]
The diff/apply sequence worked.  Here's the file I was using: 

http://share.renpy.org/the_question-1.0-linux.tar.bz2


How does permission fixing work? Do you detect shebangs and elf/macho headers? If so, I can probably just unpack the files before uploading them.

**Comment by fasterthanlime** [comment]
> Do you detect shebangs and elf/macho headers? If so, I can probably just unpack the files before uploading them.

Yup. For ELF/Mach-o, just magic numbers. Here's the code: https://github.com/itchio/wharf/blob/master/tlc/permissions.go

**Comment by fasterthanlime** [comment]
I'm probably going to add .tar.gz / .tar.bz2 support directly into butler (and the backend), but for now I'm going to find out why uploading a directory with a single .tar.bz2 / .apk didn't work for you.

By the way and if you didn't do it earlier: butler has a self-upgrade mechanism, you should be able to grab the latest version with `butler upgrade --head`. Stable versions will update too, when starting the itch app (but due to our CDN, there might be a few delays before the last version propagates).

**Comment by fasterthanlime** [comment]
Upon investigation, it looks like the patch that's pushed is 4 bytes shorter than the patch diffed locally:

```
amos at dople in ~/butler-tests/the-question
$ ls -lA *patch*
-rw-r--r--@ 1 amos  staff  16508314 Apr 29 11:16 1139-patch-default
-rw-r--r--  1 amos  staff  16508318 Apr 29 11:12 patch.pwr
```

Looks like it's easy enough to reproduce, I'll see what's happening.

**Comment by fasterthanlime** [comment]
Theory: it's the permission fixing that's breaking it.

The `GetReader` call is cached, when working on only 1 file, it's re-used between fix permissions (which reads 4 bytes) and diffing (which reads the whole file).

Usually it goes like this:

  * Open 1st file
  * Read magic number, etc.
  * Open 2nd file (trash 1st reader)
  * Read magic number, etc.
  * Permission fixed, moving on to diffing
  * Open 1st file (trash 2nd reader)
  * Start diffing ,etc.

Here it goes like this

  * Open 1st file
  * Read magic number
  * Permissions fixed, moving on
  * Open 1st file (re-use 1st reader)
  * Start diffing

TL;DR need to close after fixing permissions

**Comment by Deozaan** [comment]
I'm still unable to push a single `.apk` file. If I try to push the file itself, it seems as though butler unpacks it and repackages it all as a zip. If I put the file into a directory by itself and push the directory, butler zips the single file and I still end up with a zip file.

I don't want Android users to have to download a zip file, decompress it to get access to the apk, then launch the apk to install it. I want them to be able to just download the apk and install it.

**Comment by fasterthanlime** [comment]
I'm afraid that's still a limitation of wharf at this moment. It works on directory, either in compressed or uncompressed form. I still think uncompressing the .apk when pushing is a bug, but if it behaved correctly you'd still have a .zip on itch.io as a result, since that's the canonical archive form.

**Comment by Deozaan** [comment]
> if it behaved correctly you'd still have a .zip on itch.io as a result

Yes. It appeared on itch.io as a zip.

When I tried to push the apk file, butler said my 29MB .apk had ~90MB of data and several files/directories, which is what makes me think it unpacked it first. But in the end it was on itch.io as a ~29MB zip file.

I didn't download the zip file from itch, so I'm not sure what the unpacked contents of that file looked like (i.e., was it an apk or was it the contents of the unpacked apk?), but it did end up as a zip on the website.

**Comment by fasterthanlime** [comment]
.apk files are actually just .zip files (they're the exact same format, they just contain meta information in the form of files inside the .zip).

Since butler lets you push either an uncompressed folder, or a compressed folder (a .zip), when you try to push an .apk file, it treats it as a compressed folder - and so, you end up with a .zip on itch.io that contains the contents of the .apk!

The correct behavior would be, butler recognizes that it's not actually a .zip file (by checking the extension), and create a container with a single file in it (the .apk). The result would be a .zip file available for download on itch.io that contains a single file, the .apk.

As I said before, that butler assumes you have a "folder" and that you want a .zip to be availble for download is just a limitation we're working with right now. I am in the process of making other significant changes to the backend (esp. for large games), but supporting the "single-file" usecase is not part of it — for now.

**Comment by Deozaan** [comment]
That all makes sense for every platform except Android.

It doesn't make sense to have a .zip with an .apk inside. As you said, that's essentially zipping a .zip. In my opinion, it would be ideal when an .apk is the only file being pushed, for it to be a special case that doesn't modify or put the .apk into another .zip container.

**Comment by fasterthanlime** [comment]
> it would be ideal when an .apk is the only file being pushed, for it to be a special case that doesn't modify or put the .apk into another .zip container.

I agree! And I'll look into this special case in the future :)

**Comment by anissen** [comment]
This issue makes it impossible to use `butler` for pushing Android builds. For now I run my upload script for uploading my Web and Mac builds, but have to manually upload my Android build to avoid it turning into a Zip-file. It would be very nice to get this fixed 👍

**Comment by fasterthanlime** [comment]
It is now possible to push arbitrary files with butler. [The docs](https://itch.io/docs/butler/single-files.html) have been updated accordingly.

If you push a single compressed file, you get none of the benefits of patching, but you can automate it (since butler is a CLI tool).

.jar and .apk files are no longer mistaken as "zipped folders", they're just considered single files.

Single files are served directly to folks who download on the website, they're not wrapped in a .zip.

🎉

**Comment by renpytom** [comment]
Thank you.
